ethereum/blockwatch/block_watcher.go

```diff
@@ -302,14 +306,24 @@ func (w *Watcher) SyncToLatestBlock() error {
 		return syncErr
 	}
 	if w.shouldRevertChanges(lastStoredHeader, allEvents) {
-		// TODO(albrow): Take another look at this. Maybe extract to method.
-		if err := w.stack.Clear(); err != nil {
+		if err := w.stack.Reset(checkpoint); err != nil {
+			return err
+		}
+	} else {
+		_, err := w.stack.Checkpoint()
+		if err != nil {
 			return err
 		}
-		if _, _, err := w.stack.db.AddMiniHeaders(existingMiniHeaders); err != nil {
+		if _, err := w.db.DeleteMiniHeaders(nil); err != nil {
+			return err
+		}
+		newMiniheaders, err := w.stack.PeekAll()
+		if err != nil {
+			return err
+		}
+		if _, _, err := w.db.AddMiniHeaders(newMiniheaders); err != nil {
```

Comment:
I believe we also need to update the database at the end of `FastSyncToLatestBlock` too. Maybe I'm just missing something but it looks like we update the in-memory stack without updating the database.